pydantic_ai_slim/pydantic_ai/ui/ag_ui/_messages.py

```diff
@@ -0,0 +1,319 @@
+                                    args=tool_call.function.arguments,
+                                )
+                            )
+            else:
+                tool_call_id = msg.tool_call_id
+                tool_name = tool_calls.get(tool_call_id)
+                if tool_name is None:  # pragma: no cover
+                    raise ValueError(f'Tool call with ID {tool_call_id} not found in the history.')
+                _, provider_name, tool_call_id = tool_call_id.split('|', 2)
+
+                builder.add(
+                    BuiltinToolReturnPart(
+                        tool_name=tool_name,
+                        content=msg.content,
+                        tool_call_id=tool_call_id,
+                        provider_name=provider_name,
+                    )
+                )
+
+    # Store AG-UI message IDs in metadata for lossless roundtrip
+    result = builder.messages
+    if result:
+        _store_message_ids(result, messages)
+
+    return result
+
+
+def _store_message_ids(pydantic_messages: list[ModelMessage], ag_ui_messages: Sequence[Message]) -> None:
+    """Store original AG-UI message IDs in Pydantic AI message metadata."""
+    ag_ui_index = 0
+
+    for pydantic_msg in pydantic_messages:
+        ids_for_this_message: list[str] = []
+        parts_count = len(pydantic_msg.parts)
+
+        # Collect IDs for all parts in this message
+        for _ in range(parts_count):
+            if ag_ui_index < len(ag_ui_messages):
+                ids_for_this_message.append(ag_ui_messages[ag_ui_index].id)
+                ag_ui_index += 1
+
+        # Store IDs in metadata
+        if ids_for_this_message:
+            if pydantic_msg.metadata is None:
+                pydantic_msg.metadata = {}
+            pydantic_msg.metadata['ag_ui_message_ids'] = ids_for_this_message
+
+
+def messages_to_ag_ui(messages: Sequence[ModelMessage]) -> list[Message]:
+    """Convert Pydantic AI messages to AG-UI messages.
+
+    This utility function allows you to convert Pydantic AI messages
+    to AG-UI protocol messages.
+
+    If the Pydantic AI messages contain AG-UI message IDs in their metadata
+    (under the `'ag_ui_message_ids'` key), those IDs will be restored. Otherwise,
+    new sequential IDs will be generated.
+
+    Args:
+        messages: Sequence of Pydantic AI [`ModelMessage`][pydantic_ai.messages.ModelMessage] objects.
+
+    Returns:
+        List of AG-UI protocol messages.
+
+    Raises:
+        ValueError: If unsupported message parts are encountered.
+    """
+    ag_ui_messages: list[Message] = []
+    message_id_counter = 1
+
+    for model_msg in messages:
+        # Extract stored AG-UI message IDs from metadata if available
+        stored_ids: list[str] | None = None
+        if model_msg.metadata and 'ag_ui_message_ids' in model_msg.metadata:
+            stored_ids = model_msg.metadata['ag_ui_message_ids']
+
+        id_index = 0
+
+        def get_next_id() -> str:
```

Comment:
This is unfortunately more awkward than it would ideally be because `metadata` isn't universally available on all "MessageParts" and a `ModelRequst/Response` may contain multiple `AG_UI Messages`

@DouweM feels like `metadata` should maybe just be a base property on `MessagePart`?

Reply:
Note that we now have `provider_details` field in a bunch of places that we can use (and add where we don't have it yet).